### ISSUE TITLE
Entering Alarms in khal

### ISSUE THREAD

**Issue by petr-fischer** [pimutils/khal#301]
It will be nice If I can enter event alarms in khal UI. Something like "1 hour before", "2 hours...", "1 day before".... "1 week before" etc...

This enhancement is not about triggering alarms in the console, but about entering alarm in event edit screen. 

When synced to the phone (via CalDAV), alarms will be triggered on the phone.

**Comment by geier** [comment]
Sure, this would be very doable. PRs are very welcome.

**Comment by languitar** [comment]
I'd also be interested in that.

**Comment by languitar** [comment]
I have started to implement something here: https://github.com/languitar/khal/tree/alarms

However, when I save an event with a changed date for an alarm I end up in this error and I cannot find out what is causing this:
```
Traceback (most recent call last):
  File "/homes/jwienke/workspace/src/khal/khal/ui/__init__.py", line 756, in start_pane
    loop.run()
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/main_loop.py", line 278, in run
    self._run()
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/main_loop.py", line 376, in _run
    self.event_loop.run()
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/main_loop.py", line 682, in run
    self._loop()
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/main_loop.py", line 719, in _loop
    self._watch_files[fd]()
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/raw_display.py", line 393, in <lambda>
    event_loop, callback, self.get_available_raw_input())
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/raw_display.py", line 493, in parse_input
    callback(processed, processed_codes)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/main_loop.py", line 403, in _update
    self.process_input(keys)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/main_loop.py", line 503, in process_input
    k = self._topmost_widget.keypress(self.screen_size, k)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/wimp.py", line 643, in keypress
    return self._current_widget.keypress(size, key)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/container.py", line 1128, in keypress
    return self.body.keypress( (maxcol, remaining), key )
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/container.py", line 2269, in keypress
    key = w.keypress((mc,) + size[1:], key)
  File "/homes/jwienke/workspace/src/khal/khal/ui/__init__.py", line 639, in keypress
    return super(EventEditor, self).keypress(size, key)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/listbox.py", line 985, in keypress
    key = focus_widget.keypress((maxcol,),key)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/wimp.py", line 535, in keypress
    self._emit('click')
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/widget.py", line 463, in _emit
    signals.emit_signal(self, name, self, *args)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/signals.py", line 264, in emit
    result |= self._call_callback(callback, user_arg, user_args, args)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/urwid/signals.py", line 294, in _call_callback
    return bool(callback(*args_to_pass))
  File "/homes/jwienke/workspace/src/khal/khal/ui/__init__.py", line 625, in save
    self.collection.update(self.event)
  File "/homes/jwienke/workspace/src/khal/khal/khalendar/khalendar.py", line 346, in update
    self._calnames[event.calendar].update(event)
  File "/homes/jwienke/workspace/src/khal/khal/khalendar/khalendar.py", line 138, in update
    event.etag = self._storage.update(event.href, event, event.etag)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/vdirsyncer/storage/base.py", line 17, in inner
    return f(self, *args, **kwargs)
  File "/homes/jwienke/local/all/opt/khal/lib/python3.4/site-packages/vdirsyncer/storage/filesystem.py", line 152, in update
    raise exceptions.WrongEtagError(etag, actual_etag)
vdirsyncer.exceptions.WrongEtagError: ('1452086752403560960.000000000', '1452096282126059776.000000000')
```

Any idea how I can resolve this?

**Comment by geier** [comment]
Please remove your khal.db and try again, I believe I fixed this bug some time ago. And please tell if you need any other help.

**Comment by languitar** [comment]
I think this was caused by a backup ics file (with a different ending) in the calendar with which I was testing things. After removing the backup this doesn't happen anymore.

I think I have implemented an (ugly) UI to handle relative reminders. Maybe you could have a look at the UI. There is probably a better way to arrange these things, but I can't find something reasonable.

One thing that is missing is the input validation for the alarms triggers. So far, I do not understand how this really works for the start end editor. Could you give me some hints?

**Comment by geier** [comment]
I'm just looking at StartEndEditor and why this looks so complicated... In the end, when trying to get the new start and end dates via .newend and .newstart the validation happens and if it fails the background colors get set. This isn't optimal at all (would be much better if validation happend once the EditBox is no longer selected).

You could either copy this approach or create a Widget where you can only enter numbers.

Also, please open a PR, IMO it makes discussions easier and the code is easier to reach.

**Comment by geier** [comment]
fixed with #321
